src/compiler/crystal/semantic/cleanup_transformer.cr

```diff
@@ -997,9 +997,17 @@ module Crystal
     end
 
     def transform(node : Primitive)
-      if extra = node.extra
-        node.extra = extra.transform(self)
+      extra = node.extra
+      extra = extra.transform(self) if extra
+
+      # For `allocate` on a virtual abstract type we make `extra`
+      # be a call to `raise` at runtime. Here we just replace the
+      # "allocate" primitive with that raise call.
+      if node.name == "allocate" && extra
+        return extra
       end
+
+      node.extra = extra
```

Comment:
Just a small suggestion, allowing to skip couple of `extra` checks.

```suggestion
      if extra = node.extra
        extra = extra.transform(self)

        # For `allocate` on a virtual abstract type we make `extra`
        # be a call to `raise` at runtime. Here we just replace the
        # "allocate" primitive with that raise call.
        return extra if node.name == "allocate"

        node.extra = extra
      end
```

Reply:
Thanks! I actually thought about writing it like that too, but it makes it harder to understand. This is mixing the logic of "is there an extra" with "is this an allocate".

With how it's done right now, extra is first processed if it's there. Then, separately, we check if it's an "allocate" primitive with an extra. We check "extra" twice, but it's just a coincidence.

Reply:
Oh, I thought of it as if the existence of `extra` would create a condition, upon which a separate path is taken - consisting of 3 operations, from which, the 2nd one happens to handle an edge-case ;)